index.js

```diff
@@ -6,71 +6,78 @@
+	opts.liveness.path = setValueOrDefault(opts.liveness.path, '/live');
+	opts.liveness.createChecker = setValueOrDefault(opts.liveness.createChecker, function () { return opts.liveness.checker; });
+	opts.liveness.checker = setValueOrDefault(opts.liveness.checker, function (next) { return next(); });
+	opts.liveness.checkerTimeoutMs = setValueOrDefault(opts.liveness.checkerTimeoutMs, 20000);
+};
+
+function initChecker(checkersInfo, servicePath, checker, timeoutMs) {
+	checkersInfo[servicePath] = {
+		checker: checker,
+		timeoutMs: timeoutMs
+	};
+};
+
+function mustBeFunction(func) {
+	if (typeof func !== 'function') {
+		throw new Error('Can not create middleware checker or createChecker option must be a function');
 	}
+};
+
+function writeResponse(res, state, code) {
+	const resHeader = {
+		'Content-Type': 'application/json; charset=utf-8'
+	};
+	const content = {
+		state,
+		uptime: process.uptime(),
+		timestamp: Date.now()
+	};
+	res.writeHead(code, resHeader);
+	res.end(JSON.stringify(content, null, 2));
+};
+
+module.exports = function(opts) {
+	initOptions(opts);
+
+	mustBeFunction(opts.readiness.checker);
+	mustBeFunction(opts.liveness.checker);
+
+	mustBeFunction(opts.readiness.createChecker);
+	mustBeFunction(opts.liveness.createChecker);
 
 	let state = 'down';
 	let server;
+	const checkersInfo = {};
 
 	function handler(req, res) {
 		if (req.url == opts.readiness.path || req.url == opts.liveness.path) {
-			const resHeader = {
-				'Content-Type': 'application/json; charset=utf-8'
-			};
-
-			const content = {
-				state,
-				uptime: process.uptime(),
-				timestamp: Date.now()
-			};
-
-			if (req.url == opts.readiness.path) {
-				// Readiness if the broker started successfully.
-				// https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-readiness-probes
-				const timeout = setTimeout(function() {
-					res.writeHead(503, resHeader);
-					res.end(JSON.stringify(content, null, 2));
-				}, opts.readiness.checkerTimeoutMs);
-
-				opts.readiness.checker(function(error) {
-					clearTimeout(timeout);
-					res.writeHead((typeof error === 'undefined' && state != 'down') ? 200 : 503, resHeader);
-					res.end(JSON.stringify(content, null, 2));
-				});
-			} else {
-				// Liveness if the broker is not stopped.
-				// https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-a-liveness-command
-				const timeout = setTimeout(function() {
-					res.writeHead(503, resHeader);
-					res.end(JSON.stringify(content, null, 2));
-				}, opts.liveness.checkerTimeoutMs);
-
-				opts.liveness.checker(function(error) {
-					clearTimeout(timeout);
-					res.writeHead((typeof error === 'undefined' && state != 'down') ? 200 : 503, resHeader);
-					res.end(JSON.stringify(content, null, 2));
-				});
-			}
+			const checkerInfo = checkersInfo[req.url];
```

Comment:
What if `req.url` is not a key of `checkersInfo` (or `checkersMap`)?

This means no checker is defined at the time `handler` is called, thus we should fall back to the default behavior when no checker is defined.

Reply:
Fixed in c74b1f2